File: ios/NIReader/Models/VerificationResult.swift
import Foundation

public enum VerificationStatus: String, Codable {
    case pass = "PASS"
    case failed = "FAILED"
    case skipped = "SKIPPED"
    case warning = "WARNING"
}

public struct VerificationReport: Codable {
    public let ocrStatus: VerificationStatus
    public let nfcStatus: VerificationStatus
    public let matchingStatus: VerificationStatus
    public let overallStatus: VerificationStatus
    public let fieldChecks: [FieldMatchCheck]
    public let failureReasons: [String]
    
    public init(
        ocrStatus: VerificationStatus,
        nfcStatus: VerificationStatus,
        matchingStatus: VerificationStatus,
        overallStatus: VerificationStatus,
        fieldChecks: [FieldMatchCheck],
        failureReasons: [String] = []
    ) {
        self.ocrStatus = ocrStatus
        self.nfcStatus = nfcStatus
        self.matchingStatus = matchingStatus
        self.overallStatus = overallStatus
        self.fieldChecks = fieldChecks
        self.failureReasons = failureReasons
    }
}

public struct FieldMatchCheck: Codable {
    public let fieldName: String
    public let ocrValue: String
    public let nfcValue: String
    public let isMatch: Bool
    public let similarityScore: Float
    
    public init(fieldName: String, ocrValue: String, nfcValue: String, isMatch: Bool, similarityScore: Float = 1.0) {
        self.fieldName = fieldName
        self.ocrValue = ocrValue
        self.nfcValue = nfcValue
        self.isMatch = isMatch
        self.similarityScore = similarityScore
    }
}

// MARK: - Desktop API Responses
public struct DesktopApiResponse: Codable {
    public let status: String
    public let message: String?
    public let transactionId: String?
}

public struct HealthCheckResponse: Codable {
    public let status: String
    public let service: String?
    public let timestamp: Int64?
}

File: flutter_ni_reader/lib/data/models/verification_result.dart
enum VerificationStatus {
  PASS,
  FAILED,
  SKIPPED,
  WARNING;

  String get displayName {
    switch (this) {
      case VerificationStatus.PASS:
        return 'مطابق بنجاح (PASS)';
      case VerificationStatus.FAILED:
        return 'غير مطابق (FAILED)';
      case VerificationStatus.WARNING:
        return 'تحذير (WARNING)';
      case VerificationStatus.SKIPPED:
        return 'تم التخطي (SKIPPED)';
    }
  }
}

class FieldMatchCheck {
  final String fieldName;
  final String ocrValue;
  final String nfcValue;
  final bool isMatch;
  final double similarityScore;

  FieldMatchCheck({
    required this.fieldName,
    required this.ocrValue,
    required this.nfcValue,
    required this.isMatch,
    double? similarityScore,
  }) : similarityScore = similarityScore ?? (isMatch ? 1.0 : 0.0);

  Map<String, dynamic> toJson() => {
    'fieldName': fieldName,
    'ocrValue': ocrValue,
    'nfcValue': nfcValue,
    'isMatch': isMatch,
    'similarityScore': similarityScore,
  };

  factory FieldMatchCheck.fromJson(Map<String, dynamic> json) => FieldMatchCheck(
    fieldName: json['fieldName'] ?? '',
    ocrValue: json['ocrValue'] ?? '',
    nfcValue: json['nfcValue'] ?? '',
    isMatch: json['isMatch'] ?? false,
    similarityScore: (json['similarityScore'] as num?)?.toDouble() ?? 0.0,
  );
}

class VerificationReport {
  final VerificationStatus ocrStatus;
  final VerificationStatus nfcStatus;
  final VerificationStatus matchingStatus;
  final VerificationStatus overallStatus;
  final List<FieldMatchCheck> fieldChecks;
  final List<String> failureReasons;

  VerificationReport({
    required this.ocrStatus,
    required this.nfcStatus,
    required this.matchingStatus,
    required this.overallStatus,
    required this.fieldChecks,
    this.failureReasons = const [],
  });

  Map<String, dynamic> toJson() => {
    'ocrStatus': ocrStatus.name,
    'nfcStatus': nfcStatus.name,
    'matchingStatus': matchingStatus.name,
    'overallStatus': overallStatus.name,
    'fieldChecks': fieldChecks.map((c) => c.toJson()).toList(),
    'failureReasons': failureReasons,
  };

  factory VerificationReport.fromJson(Map<String, dynamic> json) => VerificationReport(
    ocrStatus: VerificationStatus.values.firstWhere(
      (e) => e.name == json['ocrStatus'],
      orElse: () => VerificationStatus.FAILED,
    ),
    nfcStatus: VerificationStatus.values.firstWhere(
      (e) => e.name == json['nfcStatus'],
      orElse: () => VerificationStatus.FAILED,
    ),
    matchingStatus: VerificationStatus.values.firstWhere(
      (e) => e.name == json['matchingStatus'],
      orElse: () => VerificationStatus.FAILED,
    ),
    overallStatus: VerificationStatus.values.firstWhere(
      (e) => e.name == json['overallStatus'],
      orElse: () => VerificationStatus.FAILED,
    ),
    fieldChecks: (json['fieldChecks'] as List<dynamic>?)
            ?.map((c) => FieldMatchCheck.fromJson(c))
            .toList() ??
        [],
    failureReasons: List<String>.from(json['failureReasons'] ?? []),
  );
}
